### PR TITLE
#166979383 allow admin and moderator to block reported Articles

### PR DESCRIPTION
**Pull request by Ramadhan0**
### What does this PR do?
##### Have this point working 
* FETCH ALL REPORTED ARTICLES      
    ```/api/admin/articles/reported```: Enable Admin or moderator to fetch all reported articles
* FETCH  SINGLE REPORTED ARTICLE
    ```/api/admin/article/reported/:slug```: Enable Admin or moderator to fetch reported article
*  BLOCK A REPORTED ARTICLE     
    ```/api/admin/article/block/:slug```: Enable Admiin or moderator to block reported article
* UNBLOCK A REPORTED ARTICLE           
    ```/api/admin/users/block```: Enable admin or moderator to unblock article
#### How should this be manually tested?
* Clone the repo
* Install development dependencies by type ```npm install```
* Run the server with this command ```npm run dev```

#### What are the relevant pivotal tracker stories?
[166979383](https://www.pivotaltracker.com/story/show/166979383)

### CODE REVIEW

File: src/api/controllers/adminPermissionsController.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+import generatePagination from '../../helpers/generate.pagination.details';
+
+const {
+  Article, Report, User, Reason
+} = models;
+
+/**
+ *
+ *
+ * this is the article controller
```
Comment (MCFrank16) [inline src/api/controllers/adminPermissionsController.js]: this JSDOC should be more descriptive

File: src/api/controllers/adminPermissionsController.js [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+   * @returns {array} response body
+   * @memberof getArticles
+   */
+  static async getReportedArticles(req, res) {
+    const defaultOffset = 0;
+    const defaultLimit = 10;
+    const offset = req.query.offset || defaultOffset;
+    const limit = req.query.limit || defaultLimit;
+    const paginatedArticle = await Report.findAll({
+      attributes: [['createdAt', 'ReportedOn']],
```
Comment (MCFrank16) [inline src/api/controllers/adminPermissionsController.js]: please include in a comment to explain more on how this attribute include will work

File: src/api/controllers/adminPermissionsController.js [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+  /**
+   * fetch all reported Articles
+   *
+   * @static
+   * @param {array} req - request array
+   * @param {array} res - respond array
+   * @returns {array} response body
+   * @memberof getArticles
+   */
+  static async getReportedArticles(req, res) {
```
Comment (MCFrank16) [inline src/api/controllers/adminPermissionsController.js:224]: don't you think ``getAllReportedArticles`` will be better as a class method name

File: src/api/controllers/adminPermissionsController.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+      paginationDetails: generatePagination(paginatedArticle.length,
+        paginatedArticle,
+        offset,
+        limit),
+      data: paginatedArticle
+    });
+  }
+
+  /**
+   * fetch all reported Articles
```
Comment (MCFrank16) [inline src/api/controllers/adminPermissionsController.js]: is it all reported articles or a single article

File: src/api/controllers/articleController.js [hunk 5]
```diff
@@ -138,17 +149,29 @@ export default class ArticleController {
...
+        {
           model: User,
-          attributes: ['firstName', 'lastName', 'profileImage'],
-        }]
+          attributes: ['firstName', 'lastName', 'profileImage']
+        }
+      ]
     });
 
     if (article) {
```
Comment (MCFrank16) [inline src/api/controllers/articleController.js:168]: this if statement is empty?? why??

File: src/helpers/constants/error.messages.js [hunk 6]
```diff
@@ -30,6 +30,7 @@ export default {
   ratingsNotFound: 'No ratings found for this article',
   notOwner: 'Please you must be the owner of this Article in order to modify it, Thanks',
   noArticles: 'Article not found!',
-  noMoreArticle: 'There no more articles at the moment.',
+  noMoreArticle: 'There is no more articles at the moment.',
```
Comment (MCFrank16) [inline src/helpers/constants/error.messages.js]: ``There are no more articles at the moment``

File: src/helpers/constants/error.messages.js [hunk 7]
```diff
@@ -30,6 +30,7 @@ export default {
...
   reason: 'a reason can not be a string, Please provide a number starting from 1, Thanks',
+  notMOdulator: 'you must be Moderator to perform this operation'
```
Comment (MCFrank16) [inline src/helpers/constants/error.messages.js]: ``notModerator: you must be a moderator to perform this operation``

File: src/middlewares/CheckIfModerator.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    const Moderator = await User.findOne({
+      where: {
+        email,
+        role: 'moderator'
+      }
+    });
+
+    return admin || Moderator
+      ? next()
+      : res
```
Comment (MCFrank16) [inline src/middlewares/CheckIfModerator.js]: use the error sender helper

File: src/middlewares/isArticleBlocked.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    const result = await Article.findOne({
+      where: {
+        slug,
+        isBlocked: false
+      }
+    });
+
+    return result
+      ? next()
+      : res.status(status.NOT_FOUND).json({ message: `${slug} is already blocked` });
```
Comment (MCFrank16) [inline src/middlewares/isArticleBlocked.js]: even here use the error sender

File: tests/admin.article.test.js [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+  Category: { name: 'LOVE' },
+  User: {
+    firstName: 'Alain',
+    lastName: 'Burindi',
+    profileImage: 'defaultAvatar.jpg'
+  }
+};
+
+describe('Article', () => {
+  it('should login the moderetor to get the token', (done) => {
```
Comment (MCFrank16) [inline tests/admin.article.test.js:32]: use the same method as you used up there when you were trying to get the token of the fakeAdmin

**Review by MCFrank16 — CHANGES_REQUESTED** [review]

File: src/api/controllers/adminPermissionsController.js (hunk 2 above)
Comment (Ramadhan0) [reply src/api/controllers/adminPermissionsController.js]: I am renaming createdAt to ReportedOn. that's because its the date the article reported

File: src/api/controllers/adminPermissionsController.js (hunk 3 above)
Comment (Ramadhan0) [reply src/api/controllers/adminPermissionsController.js:224]: I think that will be a long name which is not necessary because its ```articles``` not article,
@MCFrank16

File: src/api/controllers/adminPermissionsController.js (hunk 4 above)
Comment (Ramadhan0) [reply src/api/controllers/adminPermissionsController.js]: I resolved this

File: src/api/controllers/articleController.js (hunk 5 above)
Comment (Ramadhan0) [reply src/api/controllers/articleController.js:168]: this is not my work.

File: src/helpers/constants/error.messages.js (hunk 6 above)
Comment (Ramadhan0) [reply src/helpers/constants/error.messages.js]: I resolved this

File: src/middlewares/CheckIfModerator.js (hunk 8 above)
Comment (Ramadhan0) [reply src/middlewares/CheckIfModerator.js]: I resolved this too. @MCFrank16

File: tests/admin.article.test.js (hunk 10 above)
Comment (Ramadhan0) [reply tests/admin.article.test.js:32]: I have to log-in two different users, and the helper login only one user.
and the second user should be the admin or moderator @MCFrank16

File: src/api/controllers/adminPermissionsController.js (hunk 1 above)
Comment (Ramadhan0) [reply src/api/controllers/adminPermissionsController.js]: I added some comment

File: src/api/controllers/adminPermissionsController.js [hunk 11]
```diff
@@ -2,7 +2,11 @@ import models from '../models';
 import status from '../../helpers/constants/status.codes';
 import paginateUser from '../../helpers/generate.pagination.details';
 
-const { User } = models;
+const generatePagination = paginateUser;
+
+const {
+  Article, Report, User, Reason
```
Comment (MemunaHaruna) [inline src/api/controllers/adminPermissionsController.js]: @Ramadhan0 this change should already be in this file. why is it showing up as part of your changes?

File: src/api/controllers/adminPermissionsController.js [hunk 12]
```diff
@@ -62,7 +66,7 @@ export default class adminPermissions {
     const { email } = req.params;
     const { firstName, lastName } = req.body;
 
-    const result = await User.update({
+    await User.update({
```
Comment (MemunaHaruna) [inline src/api/controllers/adminPermissionsController.js:69]: @Ramadhan0 why did you change this line?

File: src/api/controllers/adminPermissionsController.js [hunk 13]
```diff
@@ -207,4 +211,155 @@ export default class adminPermissions {
...
+   * @returns {array} response body
+   * @memberof getArticles
+   */
+  static async getReportedArticles(req, res) {
+    const defaultOffset = 0;
+    const defaultLimit = 10;
+    const offset = req.query.offset || defaultOffset;
+    const limit = req.query.limit || defaultLimit;
+    const paginatedArticle = await Report.findAll({
+      attributes: [['createdAt', 'ReportedOn']],
```
Comment (MemunaHaruna) [inline src/api/controllers/adminPermissionsController.js]: @Ramadhan0 why exactly are you changing the `createdAt` attribute name? Have we been doing that across other API endpoints? Let's be consistent.

File: src/api/controllers/adminPermissionsController.js (hunk 12 above)
Comment (Ramadhan0) [reply src/api/controllers/adminPermissionsController.js:69]: this result was not being used any where @m

File: src/api/controllers/articleController.js [hunk 14]
```diff
@@ -92,17 +92,28 @@ export default class ArticleController {
     const articles = await Article.findAndCountAll({
       offset,
       limit,
-      attributes: ['id', 'title', 'description', 'body', 'slug', 'coverImage', 'tagList', 'createdAt', 'updatedAt'],
+      attributes: [
```
Comment (MemunaHaruna) [inline src/api/controllers/articleController.js:95]: @Ramadhan0 why is this showing up in your work?

Comment (Ramadhan0) [reply src/api/controllers/articleController.js:95]: this is vscode, i didn't add or remove anything @MemunaHaruna

File: src/middlewares/CheckIfModerator.js [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+   * @param {object} res the response
+   * @param { object } next the next route controller to be called
+   * @memberof CheckUser
+   * @returns {Object} res
+   */
+  static async CheckAdmins(req, res, next) {
+    const {
+      user: { email }
+    } = req.user;
+    const admin = await User.findOne({
```
Comment (MemunaHaruna) [inline src/middlewares/CheckIfModerator.js]: @Ramadhan0 why not try a Sequelize `or` here instead of making 2 separate queries?

http://docs.sequelizejs.com/manual/querying.html#where

**Review by MemunaHaruna — CHANGES_REQUESTED** [review]

File: src/api/controllers/adminPermissionsController.js [hunk 16]
```diff
@@ -207,4 +211,155 @@ export default class adminPermissions {
...
+   * @static
+   * @param {array} req - request array
+   * @param {array} res - respond array
+   * @returns {array} response body
+   * @memberof getArticles
+   */
+  static async getReportedArticle(req, res) {
+    const { slug } = req.params;
+    const reportedArticle = await Report.findOne({
+      attributes: [['createdAt', 'ReportedOn']],
```
Comment (MemunaHaruna) [inline src/api/controllers/adminPermissionsController.js]: @Ramadhan0 I don't think there's any need to fetch `createdAt` as `reportedOn`. Check for other instances of `reportedOn` and remove, both in your code and tests. Thanks.

**Review by MemunaHaruna — APPROVED** [review]
